clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/metrics/UrlMetricsInstrumentation.groovy

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.docker.registry.metrics
+
+import com.netflix.spectator.api.Id
+import com.netflix.spectator.api.Registry
+import com.squareup.okhttp.Response
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+
+import static java.lang.String.format
+
+@Component
+class UrlMetricsInstrumentation {
+
+  private Logger log = LoggerFactory.getLogger(UrlMetricsInstrumentation)
+
+  private final Registry registry
+
+  private final Id timingId
+  private final Id counterId
+
+  @Autowired
+  UrlMetricsInstrumentation(Registry registry) {
+    this.registry = registry
+    timingId = registry.createId('UrlRequestTime').withTag('className', UrlMetricsInstrumentation.simpleName)
+    counterId = registry.createId('UrlCount').withTag('className', UrlMetricsInstrumentation.simpleName)
+  }
+
+  Id getTimingId() {
+    return timingId
+  }
+
+  Id getCounterId() {
+    return counterId
+  }
+
+  void interceptorCompleted(Response response, Instant start, Instant end) {
+    Duration duration = Duration.between(start, end)
+    URL url = response.request().url()
+    registry.timer(timingId.withTag('host', url.getHost()).withTag('path', url.getPath())).record(duration.toMillis(), TimeUnit.MILLISECONDS)
```

Comment:
What are the path values? How many paths are we talking about? Regarding success, usually there is a categorization of errors so you know the types of failures. For that matter, if there is structure to the path then you might have that here instead of a flat path. And the path itself sounds suspicious since if this is an HTTP API then you can have different methods on the same path.

My concerns here are:
   1) Unbounded paths can strain the underlying metric stores and make it hard on the UX to make sense of
   2) Unstructured paths dont give much opportunity for aggregation
   3) Failures without the cause seems incomplete and forces more work to find the cause to determine if it is of interest or not

Reply:
The number of possible paths is limited by two factors:

1. The number of methods in the `DockerRegistryService` interface (See: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClient.groovy#L181 )
2. The number of `dockerRegistry.repositories` entries in your CloudDriver configuration.

So it's not unbounded, but in fact quite comparable to the number of `agent` tags in `clouddriver:executionTime__totalTime` for instance (at least in our configuration), since the `@GET("/v2/{repository}/tags/list")` is the most commonly used query for reach repository entry. (There are only `GET` methods in this interface by the way)

Example, for a `clouddriver.yaml` with a single repository:
```
curl -s http://localhost:7002/spectator/metrics | jq '.metrics["httpClient.url.count"].values[].tags | map(select(.key | contains ("path"))) | .[] .value'
"/v2/bolcom-stg-spinnaker-6b3/provisioning-image/tags/list"
"/v2/"
"/v2/"

curl -s http://localhost:7002/spectator/metrics | jq '.metrics["httpClient.url.requestTime"].values[].tags | map(select(.key | contains ("path"))) | .[] .value'
"/v2/bolcom-stg-spinnaker-6b3/provisioning-image/tags/list"
"/v2/bolcom-stg-spinnaker-6b3/provisioning-image/tags/list"
"/v2/"
"/v2/"
```

That also brings me to the *why* part of these metrics; We want to see if that `/tags/list` call is in fact the URL that gets called, and how often, and if not, which others get called instead.

We've discussed aggregating this by repository instead of by URL called, but that would mean implementing a lot of logic into the interceptor that's tightly coupled to the RestAdapter interface (which is bound to change), so we thought it best to leave it as generic as possible.
We could make an exception for the `/tags/list` call, and tag that with the registry being used?

As for the "success" part: I based the result on `response.isSuccessful()` because we're mainly interested in seeing if the client encountered an error while communicating with the registry. This could obviously easily be switched to a tag using the `response.code()` field, giving more insight on the behavior of the remote registries, but would be slightly heavier on the metric stores. 

Let me know if and what you'd like to see changed based on this input, and I'll implement those changes.

Reply:
Again, I'm not familiar with the API or usage patterns, but looking at the interface in the link your provided has "/{path}" which is what concerns me. While you dont seem to mention it in your example, you wont be the only user and I'm not familiar with access patterns. I'm not against it if it makes sense, but I'm not in a position to know if it makes sense or not and if it doesnt, then I'd like to guard against it. So could we hold off on generalizing the metric. If you are also logging calls then you still have a means to see the details.

For tags/list you could have the number of calls with a label for the repository. I imagine that would make it easier for you to answer your questions because you can filter by repository rather than by url and be more explicit about the repository (which is more concise all around).

For success, the other APIs tend to have a true/false "success" label, a label with the individual code, and a "statusCode" label that partitions the responses into 2xx, 4xx, 5xx. Having a "statusCode" could be useful but maybe not for this particular case. You can leave out the individual codes for now if they arent needed. Easy enough to add later.

Reply:
To clarify, I'm sure having {path} instrumentation would be useful. I just wish I knew more about the API.